### DISCUSSION TITLE
Can I increase the verbosity of mapping errors?

### DISCUSSION THREAD

**Discussion by heyadamhey** [leonardoporro/Detached-Mapper#72] [Q&A]
When I'm mapping an object graph, I get the following exception
```
System.NullReferenceException: 'Object reference not set to an instance of an object.'
```
Stack Trace:
```
'_context.Map<Domain.Casino>(casino)' threw an exception of type 'System.NullReferenceException'
    Data: {System.Collections.ListDictionaryInternal}
    HResult: -2147467261
    HelpLink: null
    InnerException: null
    Message: "Object reference not set to an instance of an object."
    Source: "Detached.Mappers"
    StackTrace: "   at Detached.Mappers.NotMappedParentAnnotationHandlerExtensions.NotMapped(ITypeMember member, Boolean value)\r\n   at Detached.Mappers.Annotations.NotMappedAnnotationHandler.Apply(NotMappedAttribute annotation, MapperOptions mapperOptions, ClassType typeOptions, ClassTypeMember memberOptions)\r\n   at Detached.Mappers.Annotations.AnnotationHandler`1.Apply(Attribute annotation, MapperOptions mapperOptions, ClassType typeOptions, ClassTypeMember memberOptions)\r\n   at Detached.Mappers.Types.Class.ClassTypeFactory.Create(MapperOptions options, Type type)\r\n   at Detached.Mappers.MapperOptions.<>c__DisplayClass34_0.<GetType>b__0(Type keyType)\r\n   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)\r\n   at Detached.Mappers.MapperOptions.GetType(Type type)\r\n   at Detached.Mappers.TypeMappers.ExpressionBuilder.BuildMapSingleMemberExpression(TypePairMember memberPair, Expression sourceExpr, Expression targetExpr, Expression contextExpr)\r\n   at Det
ached.Mappers.TypeMappers.ExpressionBuilder.BuildMapAllMembersExpression(TypePair typePair, Expression sourceExpr, Expression targetExpr, Expression contextExpr, Func`3 isIncluded)\r\n   at Detached.Mappers.TypeMappers.ExpressionBuilder.BuildMapMembersExpression(TypePair typePair, Func`3 isIncluded)\r\n   at Detached.Mappers.TypeMappers.Entity.Complex.EntityTypeMapperFactory.Create(Mapper mapper, TypePair typePair)\r\n   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)\r\n   at Detached.Mappers.Mapper.GetTypeMapper(TypePair typePair)\r\n   at Detached.Mappers.Mapper.Map(Object source, Type sourceClrType, Object target, Type targetClrType, IMapContext context)\r\n   at Detached.Mappers.EntityFramework.EFMapper.Map[TEntity](DbContext dbContext, Object entityOrDTO, MapParameters parameters)\r\n   at Detached.Mappers.EntityFramework.MappingDbContextExtensions.Map[TEntity](DbContext dbContext, Object entityOrDTO, MapParameters mapParams)"
    TargetSite: {Void NotMapped(Detached.Mappers.Types.ITypeMember, Boolean)}
```

for context, in `_context.Map<Domain.Casino>(casino)` `casino` is a `Domain.Casino`, so the type is the same and therefore all properties will be a 1:1 match.
The Exception Details window doesn't provide any insight into what is causing the mapping to fail.

I experienced this on a simpler object because a nullable DateTime property was null. When I ensured the property had a value, the mapping succeeded. When I try mapping an object with a lot of properties, collections etc I get the exception, but it would be great to find a simple way to debug the breakdown in the mapping. Maybe it's trying to map a nullable to a non-nullable property, which is easy enough to fix but knowing which property to address would be a huge help

Thanks
Adam

**Comment by heyadamhey** [comment]
@leonardoporro please could you advise? I tried pulling your source code down in the hope I could step into the code as it maps, but I experience the exact same result

I used the following mapping config your documentation suggests:
```
                optionsBuilder.UseMapping(mapping =>
                {
                    mapping.Default(cfg => { });

                    mapping.AddProfile(MappingProfiles.Create, cfg =>
                    {
                        cfg.Type<Entity>()
                           .FromType<DataTransferObject>()
                           .Member(u => u.DateUpdated)
                           .FromValue((u, c) => (DateTime?)DateTime.Now);
                    });

                    mapping.AddProfile(MappingProfiles.Update, cfg =>
                    {
                        cfg.Type<Entity>()
                           .FromType<Entity>()
                           .Member(u => u.DateUpdated)
                           .FromValue((u, c) => (DateTime?)DateTime.Now);
                    });

                });
```

**Comment by leonardoporro** [comment]
Hi @heyadamhey, 
Sorry for the delay, I'm sick, I've been in bed for days.
Of course I can fix whatever is broken!
Would you please put that casino entity and the configuration all together and send it to me for debugging? 
If you don't want to expose your code on the web please use my email and send it as an attachment.
Sorry again for the delay, please don't give up

**Reply by heyadamhey** [reply]
no problem at all! Thanks for getting back to me, I hope you're feeling better
I'll send you an email with the code because it's quite a complex object

**Comment by leonardoporro** [comment]
Property mapping is dynamic code and probably the most difficult part of the library, `https://github.com/leonardoporro/Detached-Mapper/blob/6e52489033d9a783aa8f91d7848b9ee27641c672/src/Detached.Mappers/TypeMappers/POCO/Complex/ComplexTypeMapperFactory.cs`
I guess I can generate a try/catch around each property  and print affected entity values

**Reply by heyadamhey** [reply]
Thanks, I'll have a look and see if I can put a breakpoint and debug without you needing to do any work on it

**Comment by leonardoporro** [comment]
do we need to add anything else here?
Please reopen if needed

**Reply by heyadamhey** [reply]
More verbose exceptions would still be very useful. When using the nuget package, an exception can be almost impossible to debug. I have switched to cloning the project and including it in my solution so I can debug. This makes it possible to track down the issue but is time-consuming to figure out which property is failing and why
